src/django/api/facility_type_processing_type.py

```diff
@@ -0,0 +1,397 @@
+    PRINTING_PROCESSING: 'Printing, Product Dyeing and Laundering',
+    ASSEMBLY_PROCESSING: 'Final Product Assembly',
+    WAREHOUSING_PROCESSING: 'Warehousing / Distribution',
+    OFFICE_PROCESSING: 'Office / HQ'
+}
+
+FACILITY_PROCESSING_TYPES = {
+    RAW_MATERIAL_PROCESSING: RAW_MATERIAL_PROCESSING_TYPES,
+    TEXTILE_PROCESSING: TEXTILE_PROCESSING_TYPES,
+    PRINTING_PROCESSING: PRINTING_PROCESSING_TYPES,
+    ASSEMBLY_PROCESSING: ASSEMBLY_PROCESSING_TYPES,
+    WAREHOUSING_PROCESSING: WAREHOUSING_PROCESSING_TYPES,
+    OFFICE_PROCESSING: OFFICE_PROCESSING_TYPES,
+}
+
+# Create a look-up of processing type -> facility type for
+# every processing type
+PROCESSING_TYPES_TO_FACILITY_TYPES = {}
+for facility_type, processing_types in FACILITY_PROCESSING_TYPES.items():
+    for processing_type_key, processing_type_value in processing_types.items():
+        PROCESSING_TYPES_TO_FACILITY_TYPES.setdefault(processing_type_value,
+                                                      ALL_FACILITY_TYPES[
+                                                          facility_type
+                                                      ])
+# Field types
+FACILITY_TYPE = 'FACILITY_TYPE'
+PROCESSING_TYPE = 'PROCESSING_TYPE'
+
+# Match types
+EXACT_MATCH = 'EXACT'
+ALIAS_MATCH = 'ALIAS'
+FUZZY_MATCH = 'FUZZY'
+
+ALL_PROCESSING_TYPES = {
+    **OFFICE_PROCESSING_TYPES,
+    **WAREHOUSING_PROCESSING_TYPES,
+    **ASSEMBLY_PROCESSING_TYPES,
+    **PRINTING_PROCESSING_TYPES,
+    **TEXTILE_PROCESSING_TYPES,
+    **RAW_MATERIAL_PROCESSING_TYPES,
+}
+
+ALL_PROCESSING_TYPES_ALIAS = {
+    **OFFICE_PROCESSING_TYPES_ALIAS,
+    **WAREHOUSING_PROCESSING_TYPES_ALIAS,
+    **ASSEMBLY_PROCESSING_TYPES_ALIAS,
+    **PRINTING_PROCESSING_TYPES_ALIAS,
+    **TEXTILE_PROCESSING_TYPES_ALIAS,
+    **RAW_MATERIAL_PROCESSING_TYPES_ALIAS,
+}
+
+
+def get_facility_and_processing_type(facility_or_processing_type):
+    """Attempts to match the input value to a facility or processing
+    type via various methods.
+    """
+    # Clean up input value
+    cleaned_input = clean(facility_or_processing_type)
+    # Assign a default value to field_type
+    field_type = PROCESSING_TYPE
+
+    # Try for exact match
+    processing_type = ALL_PROCESSING_TYPES.get(cleaned_input)
+    facility_type = ALL_FACILITY_TYPES.get(cleaned_input)
+    match_type = EXACT_MATCH
+
+    # Try for alias match
+    if not processing_type:
+        matched_value = ALL_PROCESSING_TYPES_ALIAS.get(cleaned_input)
+        match_type = ALIAS_MATCH
+
+        # Try for fuzzy match
+        if not matched_value or matched_value is None:
+            matched_value = process.extractOne(
+                cleaned_input,
+                ALL_PROCESSING_TYPES.keys()
+            )
+            match_type = FUZZY_MATCH
+
+            # Match must score 85 or higher to be considered usable.
+            if not matched_value or matched_value[1] < 85:
+                return None
```

Comment:
We may want to consider returning `(None, None, None, None)` here so that the "shape" of the return values from the function is always the same, which may make the calling code a litter clearer. 
```
>>> (a, b, c, d) = get_facility_and_processing_type('WILL NOT MATCH ANYTHING')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
TypeError: cannot unpack non-iterable NoneType object
```

Reply:
I had that at first, but changed it to `None` because `(None, None, None, None)` evaluates to `not None`. But as you point out, I have some questions about the client code implementation that may require changing this.

Reply:
Updated this. Thanks.